open_seq2seq/decoders/convs2s_decoder.py

```diff
@@ -0,0 +1,318 @@
+
+      with tf.variable_scope("layer_%d" % i):
+        if linear_proj is not None:
+          res_inputs = linear_proj(outputs)
+        else:
+          res_inputs = outputs
+
+        with tf.variable_scope("conv_layer"):
+            outputs = conv_layer(outputs)
+
+        with tf.variable_scope("attention_layer"):
+            outputs = att_layer(outputs, target_embed, encoder_outputs_a, encoder_outputs_b, input_attention_bias)
+        outputs = (outputs + res_inputs) * math.sqrt(0.5)
+
+    with tf.variable_scope("linear_layer_after_cnn_layers"):
+      outputs = self.layers[-2](outputs)
+
+    if self.mode == "train":
+      outputs = tf.nn.dropout(outputs, self.params["out_dropout_keep_prob"])
+
+    with tf.variable_scope("pre_softmax_projection"):
+      if self.layers[-1] is None:
+        logits = self.embedding_softmax_layer.linear(outputs)
+      else:
+        logits = self.layers[-1](outputs)
+
+    return tf.cast(logits, dtype=tf.float32)
+    #return logits
+
+  def predict(self, encoder_outputs, encoder_outputs_b, inputs_attention_bias):
+    """Return predicted sequence."""
+    batch_size = tf.shape(encoder_outputs)[0]
+    input_length = tf.shape(encoder_outputs)[1]
+    max_decode_length = input_length + self.params["extra_decode_length"]
+
+    symbols_to_logits_fn = self._get_symbols_to_logits_fn()
+
+    # Create initial set of IDs that will be passed into symbols_to_logits_fn.
+    initial_ids = tf.zeros([batch_size], dtype=tf.int32) + self.params["GO_SYMBOL"]
+
+    cache = {}
+    # Add encoder outputs and attention bias to the cache.
+    cache["encoder_outputs"] = encoder_outputs
+    cache["encoder_outputs_b"] = encoder_outputs_b
+    if inputs_attention_bias is not None:
+        cache["inputs_attention_bias"] = inputs_attention_bias
+
+    # Use beam search to find the top beam_size sequences and scores.
+    decoded_ids, scores = beam_search.sequence_beam_search(symbols_to_logits_fn=symbols_to_logits_fn,
+                                                            initial_ids=initial_ids,
+                                                            initial_cache=cache,
+                                                            vocab_size=self.params["tgt_vocab_size"],
+                                                            beam_size=self.params["beam_size"],
+                                                            alpha=self.params["alpha"],
+                                                            max_decode_length=max_decode_length,
+                                                            eos_id=self.params["EOS_ID"])
+
+    # Get the top sequence for each batch element
+    top_decoded_ids = decoded_ids[:, 0, :]
+    top_scores = scores[:, 0]
+
+    # this isn't particularly efficient
+    logits = self.decode_pass(top_decoded_ids, encoder_outputs, encoder_outputs_b, inputs_attention_bias)
+
+    return {"logits": logits,
+            "samples": [top_decoded_ids],
```

Comment:
"samples" has been renamed to "outputs". Can you also merge your branch with current dev0.4 before merging them back to avoid conflicts?